### PR TITLE
Adding categories

### PR DESCRIPTION
**Pull request by RomanYuzviak**

### CODE REVIEW

File: src/main/java/com/example/onlinebookstore/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+@RestController
+@Validated
+@RequestMapping(value = "/api/categories")
+@RequiredArgsConstructor
+public class CategoryController {
+    private final BookService bookService;
+    private final CategoryService categoryService;
+
+    @GetMapping
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (Cayman4010) [inline src/main/java/com/example/onlinebookstore/controller/CategoryController.java:38]: As I know, you can use just "USER" here

File: src/main/java/com/example/onlinebookstore/mapper/BookMapper.java [hunk 2]
```diff
@@ -2,13 +2,29 @@
...
+import java.util.stream.Collectors;
+import org.mapstruct.AfterMapping;
 import org.mapstruct.Mapper;
+import org.mapstruct.MappingTarget;
 
 @Mapper(config = MapperConfig.class)
 public interface BookMapper {
     BookDto toDto(Book book);
 
     Book toModel(CreateBookRequestDto bookDto);
```
Comment (Cayman4010) [inline src/main/java/com/example/onlinebookstore/mapper/BookMapper.java:19]: Maybe, "toBook" is a better name for the method?

File: src/main/java/com/example/onlinebookstore/model/Category.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import lombok.Data;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Entity
+@Table(name = "categories")
+@Data
```
Comment (Cayman4010) [inline src/main/java/com/example/onlinebookstore/model/Category.java:15]: as I know it`s better to use @Getter and @Setter instead @Data, but it`s not mistake

File: src/main/java/com/example/onlinebookstore/service/impl/CategoryServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+@Service
+@RequiredArgsConstructor
+public class CategoryServiceImpl implements CategoryService {
+    private final CategoryRepository categoryRepository;
+    private final CategoryMapper categoryMapper;
+
+    @Override
+    public CategoryDto getById(Long id) {
+        return categoryMapper.toDto(categoryRepository.findById(id).orElseThrow(() ->
+                new RuntimeException("There is not any category in db by id %d"
```
Comment (Cayman4010) [inline src/main/java/com/example/onlinebookstore/service/impl/CategoryServiceImpl.java]: maybe, you should use a custom exception?

**Review by Cayman4010 — APPROVED** [review]

File: src/main/java/com/example/onlinebookstore/controller/CategoryController.java (hunk 1 above)
Comment (liliia-ponomarenko) [reply src/main/java/com/example/onlinebookstore/controller/CategoryController.java:38]: agree

File: src/main/java/com/example/onlinebookstore/dto/category/CategoryDto.java [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package com.example.onlinebookstore.dto.category;
+
+public record CategoryDto(
+        Long id,
+        String name,
+        String description) {
+}
+
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/onlinebookstore/dto/category/CategoryDto.java]:
```suggestion
```

File: src/main/java/com/example/onlinebookstore/model/Category.java (hunk 3 above)
Comment (liliia-ponomarenko) [reply src/main/java/com/example/onlinebookstore/model/Category.java:15]: in this case, it is ok. A problem can occur if the entity has a relation

File: src/main/java/com/example/onlinebookstore/repository/CategoryRepository.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package com.example.onlinebookstore.repository;
+
+import com.example.onlinebookstore.model.Category;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/onlinebookstore/repository/CategoryRepository.java:8]: Repository  annotation missed

File: src/main/java/com/example/onlinebookstore/repository/BookRepository.java [hunk 7]
```diff
@@ -1,8 +1,21 @@
...
+import java.util.List;
+import java.util.Optional;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.EntityGraph;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
 
-public interface BookRepository extends JpaRepository<Book, Long> {
+public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/onlinebookstore/repository/BookRepository.java:14]: the same

File: src/main/java/com/example/onlinebookstore/dto/book/CreateBookRequestDto.java [hunk 8]
```diff
@@ -18,4 +19,5 @@ public class CreateBookRequestDto {
     private BigDecimal price;
     private String description;
     private String coverImage;
+    private Set<Long> categoryIdSet;
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/onlinebookstore/dto/book/CreateBookRequestDto.java]: Can you create a book with categories?

File: src/main/java/com/example/onlinebookstore/service/impl/CategoryServiceImpl.java (hunk 4 above)
Comment (liliia-ponomarenko) [reply src/main/java/com/example/onlinebookstore/service/impl/CategoryServiceImpl.java]: agree. Let’s  fix it

File: src/main/java/com/example/onlinebookstore/service/impl/CategoryServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    public CategoryDto save(CreateCategoryRequestDto categoryRequestDto) {
+        return categoryMapper.toDto(categoryRepository.save(categoryMapper
+                .toCategory(categoryRequestDto)));
+
+    }
+
+    @Override
+    public CategoryDto update(Long id, CreateCategoryRequestDto categoryRequestDto) {
+        if (!categoryRepository.existsById(id)) {
+            throw new RuntimeException("There is not book in db by id %d"
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/onlinebookstore/service/impl/CategoryServiceImpl.java]: the same

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution ;)

File: src/main/java/com/example/onlinebookstore/dto/book/CreateBookRequestDto.java (hunk 8 above)
Comment (RomanYuzviak) [reply src/main/java/com/example/onlinebookstore/dto/book/CreateBookRequestDto.java]: Yep I think

Comment (liliia-ponomarenko) [reply src/main/java/com/example/onlinebookstore/dto/book/CreateBookRequestDto.java]: Let’s  check

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Let’s check if you can save the book with categories ;)

File: src/main/java/com/example/onlinebookstore/mapper/BookMapper.java [hunk 10]
```diff
@@ -6,21 +6,38 @@
...
-    void updateBook(CreateBookRequestDto dto, @MappingTarget Book book);
+    void updateBook(@MappingTarget Book book, CreateBookRequestDto dto,
+                    @Context CategoryRepository categoryRepository);
 
     BookDtoWithoutCategoryIds toDtoWithoutCategories(Book book);
 
+    @AfterMapping
+    default void setCategories(@MappingTarget Book book,
+                               CreateBookRequestDto bookDto,
+                               @Context CategoryRepository categoryRepository) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/onlinebookstore/mapper/BookMapper.java]: It’s not good practice to pass dependency to the mapping method.  better make it on the service layer

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Let’s  fix one comment

**Review by liliia-ponomarenko — APPROVED** [review]
Great! 😊

**Pull request merged**